File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    N = len(maps)
    M = len(maps[0])
    INF = float('inf')
    steps = ((0, 1), (1, 0), (0, -1), (-1, 0))
    
    
    
    q = deque()
    q.append((0, 0, 1))
    v = [[INF] * M for _ in range(N)]
    v[0][0] = 1
    while q:
        x, y, d = q.popleft()
        if x == N-1 and y == M-1:
            return d
        
        for dx, dy in steps:
            nx, ny = x+dx, y+dy
            if nx < 0 or nx >= N or ny < 0 or ny >= M or maps[nx][ny] == 0:
                continue
            if v[nx][ny] > d+1:
                v[nx][ny] = d+1
                q.append((nx, ny, d+1))
    return -1               

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let answer = 0;
    const [N, M] = [maps.length, maps[0].length];
    const STEPS = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    
    // bfs
    const Q = [[0, 0]];
    let qIdx = 0;
    const V = Array.from({length: N}, () => 
        Array.from({length: M}, () => Infinity)
    )
    V[0][0] = 1;
    
    while (Q.length > qIdx) {
        const [x, y] = Q[qIdx];
        if ((x === N-1) && (y === M-1)) break;

        qIdx++;
        for (let i=0; i<4; ++i) {
            const dx = STEPS[i][0];
            const dy = STEPS[i][1];
            
            const nx = x+dx;
            const ny = y+dy;
            
            if (nx < 0 || nx >= N || ny < 0 || ny >= M || maps[nx][ny] === 0 || V[nx][ny] <= V[x][y]+1) continue;
            V[nx][ny] = V[x][y] + 1;
            Q.push([nx, ny]);
        }
    }
    return (V[N-1][M-1] === Infinity) ? -1 : V[N-1][M-1];
    
}